### PR TITLE
SWATCH-1983 Contract upgrades do not perform capacity reconciliation

### PR DESCRIPTION
**Pull request by kartikshahc**
Jira issue: SWATCH-1983

## Description

This PR will merge after QE approval to add the test coverage. Since the PR Merged
[SWATCH-2083: Update contract ingestion to require less data](https://github.com/RedHatInsights/rhsm-subscriptions/pull/2994#top) is already taking care of many issues the contract and subscription was getting saved. Hence there is no code change required but, QE can test the steps to close the ticket.

This problem arises when we follow a specific order of action. This occurs when we initially receive a contract for a specific date. When the UMB subscription message arrives and syncs the subscription, that is the second step. The final phase is receiving an updated contract with a new start date. I'll enumerate the output that I recorded both before and after the correction. When you run the steps on the main branch, you'll notice that the capacity is zero because no new measurements or subscriptions are created when a update contract comes in. It used to simply cancel an active subscription.

## Testing

Test the steps using wiremock.

### Setup

1.  **Wiremock entitlement gateway & subscription api**
```podman run -it --rm -p 8102:8080 --name wiremock -v $PWD/stub:/home/wiremock:z wiremock/wiremock:2.32.0 --verbose```

**Stub call for Entitlement Gateway:**
```
cat > stub/__files/swatch-1983-entitlement-gateway.json <<EOF
{
    "content": [
        {
            "rhAccountId": "7237138",
            "sourcePartner": "aws_marketplace",
            "partnerIdentities": {
                "awsCustomerId": "EK57ooq39qs",
                "customerAwsAccountId": "568056954830",
                "sellerAccountId": "568056954830"
            },
            "rhEntitlements": [
                {
                    "sku": "MW02393",
                    "subscriptionNumber": "123456"
                }
            ],
            "purchase": {
                "vendorProductCode": "1234567890abcdefghijklmno",
                "contracts": [
                    {
                        "startDate": "2023-04-21T12:29:48.569Z",
                        "dimensions": [
                            {
                                "name": "four_vcpu_hour",
                                "value": "730"
                            }
                        ]
                    }
                ]
            },
            "status": "STATUS",
            "entitlementDates": {
                "startDate": "2023-04-21T12:29:48.569Z"
            }
        }
    ],
    "page": {
        "size": 0,
        "totalElements": 1,
        "totalPages": 0,
        "number": 0
    }
}
EOF
```

```curl -X POST --data '{ "priority": 1, "request": { "urlPath": "/mock/partnerApi/v1/partnerSubscriptions", "method": "POST" }, "response": { "status": 200, "bodyFileName": "swatch-1983-entitlement-gateway.json", "headers":{"Content-Type":"application/json"} }}' http://localhost:8102/__admin/mappings/new```

**Stub call for Subscription API:**
```
cat > stub/__files/swatch-1983-subscription-api.json <<EOF
[
    {
        "id": 123456,
        "renewedId": null,
        "masterEndSystemName": "SUBSCRIPTION",
        "createdEndSystemName": "SUBSCRIPTION",
        "createdByUserName": null,
        "createdDate": 1672556530000,
        "lastUpdateEndSystemName": "SUBSCRIPTION",
        "lastUpdateUserName": null,
        "lastUpdateDate": 1672556530000,
        "externalCreatedDate": null,
        "externalLastUpdateDate": null,
        "installBaseStartDate": 1672549200000,
        "installBaseEndDate": 1704085199000,
        "inactiveDate": null,
        "oracleAccountNumber": "5456371",
        "oracleMSANumber": null,
        "oracleIBInstanceNumber": null,
        "webCustomerId": 7237138,
        "registrationNumber": null,
        "installationNumber": null,
        "subscriptionNumber": "123456",
        "quantity": 10,
        "subscriptionProducts": [
            {"sku": "MW02393"}
        ],
        "customer": {
            "id": 7237138,
            "name": "Dummy Customer",
            "password": null,
            "creditApplicationCompleted": false,
            "customerType": "organization",
            "oracleCustomerNumber": 5456371,
            "namedAccount": false,
            "createdDate": 1397548100000,
            "updatedDate": 1612591188913
        },
        "effectiveStartDate": 1672549200000,
        "effectiveEndDate": 1729998800000,
        "externalReferences": {
            "awsMarketplace": {
                "productCode": "6n58d3s3qpvk22dgew2gal7w3",
                "customerID": "EK57ooq39qs",
                "sellerAccount": "568056954830",
                "customerAccountID": "568056954830"
            }
        }
    }
]
EOF
```

```curl -X POST --data '{ "priority": 1, "request": { "method": "GET", "urlPattern": "/search/criteria;.+=123456.*" }, "response": { "status": 200, "bodyFileName": "swatch-1983-subscription-api.json", "headers": { "Content-Type": "application/json" } } }' http://localhost:8102/__admin/mappings/new ```

```curl -X POST --data '{ "priority": 1, "request": { "method": "GET", "urlPattern": "/search/criteria;.+=7237138.*" }, "response": { "status": 200, "bodyFileName": "swatch-1983-subscription-api.json", "headers": { "Content-Type": "application/json" } } }' http://localhost:8102/__admin/mappings/new```

2. Run the core and swatch-contract components:

swatch-core:
`SUBSCRIPTION_URL=http://localhost:8102/ SERVER_PORT=8101 DEV_MODE=true SPRING_PROFILES_ACTIVE=capacity-ingress,api,kafka-queue ./gradlew :bootRun`

swatch-contract:
`SUBSCRIPTION_URL=http://localhost:8102/ ENTITLEMENT_GATEWAY_URL=http://localhost:8102/mock/partnerApi ./gradlew :swatch-contracts:quarkusDev`

3. Add org by optin and also add offering record to match with sku:
`INSERT INTO public.offering (sku, product_name, product_family, cores, sockets, hypervisor_cores, hypervisor_sockets, role, sla, usage, description, has_unlimited_usage, derived_sku, metered) VALUES ('MW02393', 'OpenShift Online', 'OpenShift Enterprise', null, null, null, null, 'moa-hostedcontrolplane', 'Premium', '', 'Red Hat OpenShift Service on AWS Hosted Control Planes (Hourly)', false, '', null);`

### Steps and Verification

1. Create contract:
```
curl -X POST 'http://localhost:8000/api/swatch-contracts/internal/rpc/partner/contracts'   -H 'accept: application/json'   -H 'x-rh-swatch-psk: placeholder'   -H 'Content-Type: application/json'   -d '{
  "action": "contract-update",
  "redHatSubscriptionNumber": "123456",
  "currentDimensions": [
    {
      "dimensionName": "control_plane",
      "dimensionValue": "730",
      "expirationDate": "2024-11-29T23:59:59.999Z"
    }
  ],
  "cloudIdentifiers": {
    "awsCustomerId": "EK57ooq39qs",
    "awsCustomerAccountId": "568056954830",
    "productCode": "6n58d3s3qpvk22dgew2gal7w3"
  }
}' 
```

Response:
`{"status":"SUCCESS","message":"New contract created"}`

Verify the above code should have created entries similar to:
Subscription table:
```
MW02393,7237138,123456,1,2023-04-21 12:29:48.569000 +00:00,,1234567890abcdefghijklmno;EK57ooq39qs;568056954830,123456,aws,568056954830
```

Subscription product ids:
`rosa,123456,2023-04-21 12:29:48.569000 +00:00`

Subscription measurements:
```
123456,2023-04-21 12:29:48.569000 +00:00,Instance-hours,PHYSICAL,730
123456,2023-04-21 12:29:48.569000 +00:00,Cores,PHYSICAL,2920
```

Contract:
`0b59da1b-eb8f-4705-9939-39295251a88a,123456,2024-02-01 23:27:19.189295 +00:00,2023-04-21 12:29:48.569000 +00:00,,7237138,MW02393,aws,568056954830,rosa,6n58d3s3qpvk22dgew2gal7w3,1234567890abcdefghijklmno;EK57ooq39qs;568056954830`

Contract_metrics:
```
0b59da1b-eb8f-4705-9939-39295251a88a,control_plane,730
0b59da1b-eb8f-4705-9939-39295251a88a,four_vcpu_hour,730
```

2. Check for capacity and it should show proper results:
`http ':8101/api/rhsm-subscriptions/v1/subscriptions/products/rosa' uom==Cores x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"7237138"}}}' | base64 -w0)`

**Verify:**
```
{
    "data": [
        {
            "billing_provider": "aws",
            "capacity": 2920,
            "has_infinite_quantity": false,
            "hypervisor_capacity": 0,
            "product_name": "Red Hat OpenShift Service on AWS Hosted Control Planes (Hourly)",
            "quantity": 1,
            "service_level": "Premium",
            "sku": "MW02393",
            "subscriptions": [
                {
                    "id": "123456",
                    "number": "123456"
                }
            ],
            "total_capacity": 2920,
            "uom": "cores",
            "usage": ""
        }
    ],
    "meta": {
        "count": 1,
        "product": "rosa",
        "subscription_type": "On-demand",
        "uom": "cores"
    }
}
```

2. Using wiremock as UMB subscription sync subscription using:
`curl -X 'PUT'   'http://localhost:8101/api/rhsm-subscriptions/v1/internal/subscriptions/sync/org/7237138'   -H 'accept: application/vnd.api+json'   -H 'x-rh-identity: eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6IiIsInR5cGUiOiJVc2VyIiwidXNlciI6eyJpc19vcmdfYWRtaW4iOnRydWV9LCJpbnRlcm5hbCI6eyJvcmdfaWQiOiI3MjM3MTM4In19fQ'`

Verify an entry gets created in subscription table:
`MW02393,7237138,123456,10,2023-01-01 05:00:00.000000 +00:00,2024-10-27 03:13:20.000000 +00:00,6n58d3s3qpvk22dgew2gal7w3;EK57ooq39qs;568056954830,123456,aws,568056954830`

You should still see the same capacity when you run capacity API.

3. Update contract:
Change the dimension value in `stub/__files/swatch-1983-entitlement-gateway.json` like below:
```
 "contracts": [
                    {
                        "startDate": "2023-04-21T12:29:48.569Z",
                        "dimensions": [
                            {
                                "name": "four_vcpu_hour",
                                "value": "735"
                            }
                        ]
                    }
                ] 
```

Run the stub command again to get a new mock response when Entitlement Gateway API is called:
```curl -X POST --data '{ "priority": 1, "request": { "urlPath": "/mock/partnerApi/v1/partnerSubscriptions", "method": "POST" }, "response": { "status": 200, "bodyFileName": "swatch-1983-entitlement-gateway.json", "headers":{"Content-Type":"application/json"} }}' http://localhost:8102/__admin/mappings/new```

Execute contract update:
```
curl -X POST 'http://localhost:8000/api/swatch-contracts/internal/rpc/partner/contracts'   -H 'accept: application/json'   -H 'x-rh-swatch-psk: placeholder'   -H 'Content-Type: application/json'   -d '{
  "action": "contract-update",
  "redHatSubscriptionNumber": "123456",
  "currentDimensions": [
    {
      "dimensionName": "control_plane",
      "dimensionValue": "735",
      "expirationDate": "2024-11-30T23:59:59.999Z"
    }
  ],
  "cloudIdentifiers": {
    "awsCustomerId": "EK57ooq39qs",
    "awsCustomerAccountId": "568056954830",
    "productCode": "6n58d3s3qpvk22dgew2gal7w3"
  }
}' 
```

Response: `{"status":"SUCCESS","message":"Previous contract archived and new contract created"}`

4. Run capacity API again:
`http ':8101/api/rhsm-subscriptions/v1/subscriptions/products/rosa' uom==Cores x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"7237138"}}}' | base64 -w0)`

**Verify:** 
This step on "**main**" branch will return zero capacity because it doesn't create subscription and just ends the existing subscription. Obviously it doesn't create subscription measurements since new subscription was not created:
```
{
    "data": [
        {
            "billing_provider": "aws",
            "capacity": 0,
            "has_infinite_quantity": false,
            "hypervisor_capacity": 0,
            "next_event_date": "2024-10-27T03:13:20Z",
            "next_event_type": "Subscription End",
            "product_name": "Red Hat OpenShift Service on AWS Hosted Control Planes (Hourly)",
            "quantity": 10,
            "service_level": "Premium",
            "sku": "MW02393",
            "subscriptions": [
                {
                    "id": "123456",
                    "number": "123456"
                }
            ],
            "total_capacity": 0,
            "usage": ""
        }
    ],
    "meta": {
        "count": 1,
        "product": "rosa",
        "subscription_type": "On-demand",
        "uom": "cores"
    }
}

```

This step when you perform on "**current branch**" will return correct capacity because it created subscription measurements and matches join query:
```
{
    "data": [
        {
            "billing_provider": "aws",
            "capacity": 2940,
            "has_infinite_quantity": false,
            "hypervisor_capacity": 0,
            "product_name": "Red Hat OpenShift Service on AWS Hosted Control Planes (Hourly)",
            "quantity": 1,
            "service_level": "Premium",
            "sku": "MW02393",
            "subscriptions": [
                {
                    "id": "123456",
                    "number": "123456"
                }
            ],
            "total_capacity": 2940,
            "uom": "cores",
            "usage": ""
        }
    ],
    "meta": {
        "count": 1,
        "product": "rosa",
        "subscription_type": "On-demand",
        "uom": "cores"
    }
}
```

Final results in table:
Subscription:
```
MW02393,7237138,123456,1,2023-04-21 12:29:48.569000 +00:00,2024-02-01 23:34:21.461173 +00:00,1234567890abcdefghijklmno;EK57ooq39qs;568056954830,123456,aws,568056954830
MW02393,7237138,123456,10,2023-01-01 05:00:00.000000 +00:00,2024-10-27 03:13:20.000000 +00:00,6n58d3s3qpvk22dgew2gal7w3;EK57ooq39qs;568056954830,123456,aws,568056954830
MW02393,7237138,123456,1,2024-02-01 23:34:21.461173 +00:00,,1234567890abcdefghijklmno;EK57ooq39qs;568056954830,123456,aws,568056954830
```

Subscription Product ids:
```
rosa,123456,2023-04-21 12:29:48.569000 +00:00
rosa,123456,2024-02-01 23:34:21.461173 +00:00
```

Subscription measurements:
```
123456,2023-04-21 12:29:48.569000 +00:00,Instance-hours,PHYSICAL,730
123456,2023-04-21 12:29:48.569000 +00:00,Cores,PHYSICAL,2920
123456,2024-02-01 23:34:21.461173 +00:00,Instance-hours,PHYSICAL,735
123456,2024-02-01 23:34:21.461173 +00:00,Cores,PHYSICAL,2940
```

Contracts:
```
210d5a35-41c6-4b4b-8a14-931e2dce0966,123456,2024-02-01 23:34:21.461173 +00:00,2024-02-01 23:34:21.461173 +00:00,,7237138,MW02393,aws,568056954830,rosa,6n58d3s3qpvk22dgew2gal7w3,1234567890abcdefghijklmno;EK57ooq39qs;568056954830
0b59da1b-eb8f-4705-9939-39295251a88a,123456,2024-02-01 23:34:21.461173 +00:00,2023-04-21 12:29:48.569000 +00:00,2024-02-01 23:34:21.461173 +00:00,7237138,MW02393,aws,568056954830,rosa,6n58d3s3qpvk22dgew2gal7w3,1234567890abcdefghijklmno;EK57ooq39qs;568056954830
```

Contract_metrics:
```
0b59da1b-eb8f-4705-9939-39295251a88a,control_plane,730
0b59da1b-eb8f-4705-9939-39295251a88a,four_vcpu_hour,730
210d5a35-41c6-4b4b-8a14-931e2dce0966,control_plane,735
210d5a35-41c6-4b4b-8a14-931e2dce0966,four_vcpu_hour,735
```

### CODE REVIEW

**Comment by kartikshahc** [comment]
@TrayvonMcKnight @ntkathole I will merge PR after QE approval to add the test coverage. Since the PR Merged
https://github.com/RedHatInsights/rhsm-subscriptions/pull/2994#top is already taking care of many issues when the contract and subscription were getting saved. Hence there is no code change required but, QE can test the above steps to close the ticket.

**Review by TrayvonMcKnight — APPROVED** [review]
Walked through the provided steps:
https://drive.google.com/file/d/14AUHWEWduoPVHkuv3lYmNuE5n6UoTN64/view?usp=drive_link

**Pull request merged**
